### PR TITLE
Add documentation for implicit state management

### PR DESCRIPTION
**Pull request by Tabrizian**

### CODE REVIEW

File: docs/architecture.md [hunk 1]
```diff
@@ -257,6 +257,60 @@ sequence_batching {
...
+describes the name of the output tensor produced by the model that contains
+output state. The output state provided by the model in the *i<sup>th</sup>*
+request in the sequence will be used as the input state in the
+*i+1<sup>th</sup>* request.  The *dims* field specifies the dimensions of the
+state tensors. When the *dims* field contains variable-sized dimensions, the
+shape of the input state and output state does not have to match.
+
+For debugging purposes, the client can request the output state. In order to
+allow the client to request the output state, the *output* section of the model
+configuration must list the output state as one of the model outputs. Note that
```
Comment (deadeyegoodwin) [inline docs/architecture.md:294]: Reference the "input and output" section of model_configuration.md

File: docs/architecture.md [hunk 2]
```diff
@@ -257,6 +257,60 @@ sequence_batching {
...
+
+The starting request in the sequence contains uninitialized data for the input
+state. The model can use the start flag in the request to detect the beginning
+of a new sequence and initialize the model state by providing the initial state
+in the model output. If the *dims* section in the *state* description of the
+model contains variable-sized dimensions, Triton will use *1* for every
+variable-sized dimension for the starting request. For other non-starting
+requests in the sequence, the input state is the output state of the previous
+request in the sequence.
+
```
Comment (deadeyegoodwin) [inline docs/architecture.md:313]: Point to an example or CI test that demonstrates.

**Review by deadeyegoodwin — CHANGES_REQUESTED** [review]

**Review by GuanLuo — DISMISSED** [review]

**Review by tanmayv25 — COMMENTED** [review]
lgtm

File: docs/architecture.md [hunk 3]
```diff
@@ -257,6 +257,67 @@ sequence_batching {
...
+in the model output. If the *dims* section in the *state* description of the
+model contains variable-sized dimensions, Triton will use *1* for every
+variable-sized dimension for the starting request. For other non-starting
+requests in the sequence, the input state is the output state of the previous
+request in the sequence. For an example ONNX model that uses implicit state to
+accumulate all the inputs in a sequence you can refer to [this CI
+test](../qa/common/gen_qa_implicit_models.py#L130). This model stores the
+partial sum in Triton using implicit state. For state initialization, if the
+request is starting, the model puts the "INPUT" in the "OUTPUT\_STATE" field.
+For non-starting requests, it sets the "OUTPUT\_STATE" and "OUTPUT" to "INPUT"
```
Comment (tanmayv25) [inline docs/architecture.md:318]: IMO no need to talk about OUTPUT here.
> it sets the "OUTPUT\_STATE" to "INPUT" \+  "INPUT\_STATE"

You can elaborate on the model after L313-L315 to what OUTPUT is if you want.

**Review by tanmayv25 — APPROVED** [review]

**Review by deadeyegoodwin — APPROVED** [review]

**Pull request merged**